File: go-server/internal/tools/react.go
// Package tools provides MCP tool implementations
package tools

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/oyin-bo/autoreply/go-server/internal/auth"
	"github.com/oyin-bo/autoreply/go-server/internal/mcp"
	"github.com/oyin-bo/autoreply/go-server/pkg/errors"
)

// ReactTool implements the react tool for batch post operations (like, unlike, repost, delete)
type ReactTool struct {
	credStore *auth.CredentialStore
	client    *http.Client
}

// NewReactTool creates a new react tool
func NewReactTool() (*ReactTool, error) {
	credStore, err := auth.NewCredentialStore()
	if err != nil {
		return nil, fmt.Errorf("failed to initialize credential store: %w", err)
	}

	return &ReactTool{
		credStore: credStore,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}, nil
}

// Name returns the tool name
func (t *ReactTool) Name() string {
	return "react"
}

// Description returns the tool description
func (t *ReactTool) Description() string {
	return "Perform batch reactions on posts: like, unlike, repost, and delete. Supports mixing at:// URIs and https://bsky.app/... URLs."
}

// InputSchema returns the JSON schema for tool input
func (t *ReactTool) InputSchema() mcp.InputSchema {
	return mcp.InputSchema{
		Type: "object",
		Properties: map[string]mcp.PropertySchema{
			"reactAs": {
				Type:        "string",
				Description: "Handle or DID to react as (uses default account if not specified)",
			},
			"like": {
				Type:        "string",
				Description: "Post URIs/URLs to like (comma or newline separated)",
			},
			"unlike": {
				Type:        "string",
				Description: "Post URIs/URLs to unlike (remove like) (comma or newline separated)",
			},
			"repost": {
				Type:        "string",
				Description: "Post URIs/URLs to repost (comma or newline separated)",
			},
			"delete": {
				Type:        "string",
				Description: "Post URIs/URLs to delete (must be your own posts) (comma or newline separated)",
			},
		},
	}
}

// Call executes the react tool
func (t *ReactTool) Call(ctx context.Context, args map[string]interface{}, _ *mcp.Server) (*mcp.ToolResult, error) {
	// Extract reactAs parameter (optional, uses default if not provided)
	var reactAs string
	if reactAsRaw, ok := args["reactAs"]; ok {
		if reactAsStr, ok := reactAsRaw.(string); ok {
			reactAs = strings.TrimSpace(strings.TrimPrefix(reactAsStr, "@"))
		}
	}

	// Get credentials for the user
	creds, err := t.getCredentials(reactAs)
	if err != nil {
		return nil, err
	}

	// Process all operations
	results := &OperationResults{
		Successes: []string{},
		Failures:  []OperationFailure{},
	}

	// Process like operations
	for _, uri := range parseURIList(args["like"]) {
		if err := t.likePost(ctx, creds, uri); err != nil {
			results.Failures = append(results.Failures, OperationFailure{
				Operation: "like",
				URI:       uri,
				Error:     err.Error(),
			})
		} else {
			results.Successes = append(results.Successes, fmt.Sprintf("Liked: %s", uri))
		}
	}

	// Process unlike operations
	for _, uri := range parseURIList(args["unlike"]) {
		if err := t.unlikePost(ctx, creds, uri); err != nil {
			results.Failures = append(results.Failures, OperationFailure{
				Operation: "unlike",
				URI:       uri,
				Error:     err.Error(),
			})
		} else {
			results.Successes = append(results.Successes, fmt.Sprintf("Unliked: %s", uri))
		}
	}

	// Process repost operations
	for _, uri := range parseURIList(args["repost"]) {
		if err := t.repostPost(ctx, creds, uri); err != nil {
			results.Failures = append(results.Failures, OperationFailure{
				Operation: "repost",
				URI:       uri,
				Error:     err.Error(),
			})
		} else {
			results.Successes = append(results.Successes, fmt.Sprintf("Reposted: %s", uri))
		}
	}

	// Process delete operations
	for _, uri := range parseURIList(args["delete"]) {
		if err := t.deletePost(ctx, creds, uri); err != nil {
			results.Failures = append(results.Failures, OperationFailure{
				Operation: "delete",
				URI:       uri,
				Error:     err.Error(),
			})
		} else {
			results.Successes = append(results.Successes, fmt.Sprintf("Deleted: %s", uri))
		}
	}

	// Build response message
	message := t.formatResults(results)

	return &mcp.ToolResult{
		Content: []mcp.ContentItem{
			{
				Type: "text",
				Text: message,
			},
		},
		IsError: len(results.Failures) > 0 && len(results.Successes) == 0,
	}, nil
}

// OperationResults tracks successes and failures of batch operations
type OperationResults struct {
	Successes []string
	Failures  []OperationFailure
}

// OperationFailure represents a failed operation
type OperationFailure struct {
	Operation string
	URI       string
	Error     string
}

// getCredentials retrieves credentials for a user (by handle or DID)
func (t *ReactTool) getCredentials(reactAs string) (*auth.Credentials, error) {
	// If no reactAs specified, use default
	if reactAs == "" {
		defaultHandle, err := t.credStore.GetDefault()
		if err != nil {
			return nil, errors.NewMCPError(errors.InvalidInput, "No default account set. Please login first or specify reactAs parameter.")
		}
		reactAs = defaultHandle
	}

	// Load credentials
	creds, err := t.credStore.Load(reactAs)
	if err != nil {
		return nil, errors.Wrap(err, errors.NotFound, fmt.Sprintf("No credentials found for %s. Please login first.", reactAs))
	}

	return creds, nil
}

// likePost creates a like record for a post
func (t *ReactTool) likePost(ctx context.Context, creds *auth.Credentials, postURI string) error {
	// Parse the post reference
	postRef, err := parsePostReference(postURI)
	if err != nil {
		return err
	}

	// First, get the post to get its CID
	post, err := t.getPost(ctx, creds, postRef)
	if err != nil {
		return err
	}

	// Create like record
	endpoint := "https://bsky.social/xrpc/com.atproto.repo.createRecord"

	record := map[string]interface{}{
		"$type": "app.bsky.feed.like",
		"subject": map[string]interface{}{
			"uri": post.URI,
			"cid": post.CID,
		},
		"createdAt": time.Now().UTC().Format(time.RFC3339),
	}

	requestBody := map[string]interface{}{
		"repo":       creds.DID,
		"collection": "app.bsky.feed.like",
		"record":     record,
	}

	return t.makeAuthenticatedRequest(ctx, creds, "POST", endpoint, requestBody)
}

// unlikePost removes a like from a post
func (t *ReactTool) unlikePost(ctx context.Context, creds *auth.Credentials, postURI string) error {
	// Parse the post reference
	postRef, err := parsePostReference(postURI)
	if err != nil {
		return err
	}

	// First, find the like record
	likeURI, err := t.findLikeRecord(ctx, creds, postRef)
	if err != nil {
		return err
	}

	if likeURI == "" {
		return fmt.Errorf("no like found for post %s", postURI)
	}

	// Parse the like URI to extract rkey
	parts := strings.Split(strings.TrimPrefix(likeURI, "at://"), "/")
	if len(parts) < 3 {
		return fmt.Errorf("invalid like URI format: %s", likeURI)
	}
	rkey := parts[2]

	// Delete the like record
	endpoint := "https://bsky.social/xrpc/com.atproto.repo.deleteRecord"

	requestBody := map[string]interface{}{
		"repo":       creds.DID,
		"collection": "app.bsky.feed.like",
		"rkey":       rkey,
	}

	return t.makeAuthenticatedRequest(ctx, creds, "POST", endpoint, requestBody)
}

// repostPost creates a repost record for a post
func (t *ReactTool) repostPost(ctx context.Context, creds *auth.Credentials, postURI string) error {
	// Parse the post reference
	postRef, err := parsePostReference(postURI)
	if err != nil {
		return err
	}

	// Get the post to get its CID
	post, err := t.getPost(ctx, creds, postRef)
	if err != nil {
		return err
	}

	// Create repost record
	endpoint := "https://bsky.social/xrpc/com.atproto.repo.createRecord"

	record := map[string]interface{}{
		"$type": "app.bsky.feed.repost",
		"subject": map[string]interface{}{
			"uri": post.URI,
			"cid": post.CID,
		},
		"createdAt": time.Now().UTC().Format(time.RFC3339),
	}

	requestBody := map[string]interface{}{
		"repo":       creds.DID,
		"collection": "app.bsky.feed.repost",
		"record":     record,
	}

	return t.makeAuthenticatedRequest(ctx, creds, "POST", endpoint, requestBody)
}

// deletePost deletes a post (must be user's own post)
func (t *ReactTool) deletePost(ctx context.Context, creds *auth.Credentials, postURI string) error {
	// Parse the post reference
	postRef, err := parsePostReference(postURI)
	if err != nil {
		return err
	}

	// Verify the post belongs to the user
	if postRef.DID != creds.DID {
		return fmt.Errorf("cannot delete post that doesn't belong to you")
	}

	// Delete the post record
	endpoint := "https://bsky.social/xrpc/com.atproto.repo.deleteRecord"

	requestBody := map[string]interface{}{
		"repo":       creds.DID,
		"collection": "app.bsky.feed.post",
		"rkey":       postRef.RKey,
	}

	return t.makeAuthenticatedRequest(ctx, creds, "POST", endpoint, requestBody)
}

// PostInfo represents basic post information
type PostInfo struct {
	URI string
	CID string
}

// getPost fetches a post's URI and CID
func (t *ReactTool) getPost(ctx context.Context, creds *auth.Credentials, postRef *PostReference) (*PostInfo, error) {
	endpoint := fmt.Sprintf("https://bsky.social/xrpc/com.atproto.repo.getRecord?repo=%s&collection=app.bsky.feed.post&rkey=%s",
		postRef.DID, postRef.RKey)

	req, err := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", creds.AccessToken))
	req.Header.Set("User-Agent", "autoreply/1.0")

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch post: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var errorResp map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&errorResp)
		return nil, fmt.Errorf("failed to fetch post with status %d: %v", resp.StatusCode, errorResp)
	}

	var recordResp struct {
		URI string `json:"uri"`
		CID string `json:"cid"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&recordResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &PostInfo{
		URI: recordResp.URI,
		CID: recordResp.CID,
	}, nil
}

// findLikeRecord finds the like record URI for a given post
func (t *ReactTool) findLikeRecord(ctx context.Context, creds *auth.Credentials, postRef *PostReference) (string, error) {
	// List records in the likes collection
	endpoint := fmt.Sprintf("https://bsky.social/xrpc/com.atproto.repo.listRecords?repo=%s&collection=app.bsky.feed.like&limit=100",
		creds.DID)

	req, err := http.NewRequestWithContext(ctx, "GET", endpoint, nil)
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", creds.AccessToken))
	req.Header.Set("User-Agent", "autoreply/1.0")

	resp, err := t.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to list likes: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var errorResp map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&errorResp)
		return "", fmt.Errorf("failed to list likes with status %d: %v", resp.StatusCode, errorResp)
	}

	var listResp struct {
		Records []struct {
			URI   string                 `json:"uri"`
			Value map[string]interface{} `json:"value"`
		} `json:"records"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&listResp); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	// Find the like for this specific post
	targetURI := fmt.Sprintf("at://%s/app.bsky.feed.post/%s", postRef.DID, postRef.RKey)
	for _, record := range listResp.Records {
		if subject, ok := record.Value["subject"].(map[string]interface{}); ok {
			if uri, ok := subject["uri"].(string); ok && uri == targetURI {
				return record.URI, nil
			}
		}
	}

	return "", nil
}

// makeAuthenticatedRequest makes an authenticated HTTP request
func (t *ReactTool) makeAuthenticatedRequest(ctx context.Context, creds *auth.Credentials, method, endpoint string, body interface{}) error {
	var reqBody *bytes.Buffer
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("failed to marshal request: %w", err)
		}
		reqBody = bytes.NewBuffer(jsonData)
	} else {
		reqBody = bytes.NewBuffer(nil)
	}

	req, err := http.NewRequestWithContext(ctx, method, endpoint, reqBody)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", creds.AccessToken))
	req.Header.Set("User-Agent", "autoreply/1.0")

	resp, err := t.client.Do(req)
	if err != nil {
		return fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		var errorResp map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&errorResp)
		return fmt.Errorf("request failed with status %d: %v", resp.StatusCode, errorResp)
	}

	return nil
}

// formatResults formats the operation results as markdown
func (t *ReactTool) formatResults(results *OperationResults) string {
	var sb strings.Builder

	sb.WriteString("# Reaction Operations\n\n")

	// Show successes
	if len(results.Successes) > 0 {
		sb.WriteString(fmt.Sprintf("## ✓ Successful Operations (%d)\n\n", len(results.Successes)))
		for _, success := range results.Successes {
			sb.WriteString(fmt.Sprintf("- %s\n", success))
		}
		sb.WriteString("\n")
	}

	// Show failures
	if len(results.Failures) > 0 {
		sb.WriteString(fmt.Sprintf("## ✗ Failed Operations (%d)\n\n", len(results.Failures)))
		for _, failure := range results.Failures {
			sb.WriteString(fmt.Sprintf("- **%s** `%s`: %s\n", failure.Operation, failure.URI, failure.Error))
		}
		sb.WriteString("\n")
	}

	// Summary
	total := len(results.Successes) + len(results.Failures)
	sb.WriteString(fmt.Sprintf("**Summary:** %d successful, %d failed out of %d total operations\n",
		len(results.Successes), len(results.Failures), total))

	return sb.String()
}

// parseURIList normalizes a tool argument into a list of post URIs.
// It accepts either:
// - string: comma/newline/semicolon-separated values
// - []interface{} or []string: list of strings
func parseURIList(v interface{}) []string {
	var out []string
	if v == nil {
		return out
	}
	switch vv := v.(type) {
	case string:
		// Split on comma, newline, or semicolon
		parts := strings.FieldsFunc(vv, func(r rune) bool {
			switch r {
			case ',', '\n', ';':
				return true
			default:
				return false
			}
		})
		for _, p := range parts {
			s := strings.TrimSpace(p)
			if s != "" {
				out = append(out, s)
			}
		}
	case []interface{}:
		for _, item := range vv {
			if s, ok := item.(string); ok {
				s = strings.TrimSpace(s)
				if s != "" {
					out = append(out, s)
				}
			}
		}
	case []string:
		for _, s := range vv {
			s2 := strings.TrimSpace(s)
			if s2 != "" {
				out = append(out, s2)
			}
		}
	}
	return out
}

File: rust-server/src/tools/react.rs
//! React tool implementation
//!
//! Implements the `react(reactAs, like, unlike, repost, delete)` MCP tool
//! Supports batching multiple operations in a single call

use crate::auth::storage::CredentialStorage;
use crate::auth::SessionManager;
use crate::cli::ReactArgs;
use crate::error::AppError;
use crate::mcp::{McpResponse, ToolResult};
use anyhow::Result;
use serde_json::Value;
use tokio::time::{timeout, Duration};
use tracing::debug;

/// Handle react tool call
pub async fn handle_react(id: Option<Value>, args: Value) -> McpResponse {
    // Set total timeout to 120 seconds
    match timeout(Duration::from_secs(120), handle_react_impl(args)).await {
        Ok(result) => match result {
            Ok(content) => McpResponse::success(id, serde_json::to_value(content).unwrap()),
            Err(e) => McpResponse::error(id, e.error_code(), &e.message()),
        },
        Err(_) => McpResponse::error(id, "timeout", "React request exceeded 120 second timeout"),
    }
}

async fn handle_react_impl(args: Value) -> Result<ToolResult, AppError> {
    // Parse arguments
    let react_args: ReactArgs = serde_json::from_value(args)
        .map_err(|e| AppError::InvalidInput(format!("Invalid arguments: {}", e)))?;

    // Execute using shared implementation
    execute_react(react_args).await
}

/// Execute react tool (shared implementation for MCP and CLI)
pub async fn execute_react(react_args: ReactArgs) -> Result<ToolResult, AppError> {
    debug!(
        "React request for account: {}, like: {}, unlike: {}, repost: {}, delete: {}",
        react_args.reactAs,
        react_args.like.len(),
        react_args.unlike.len(),
        react_args.repost.len(),
        react_args.delete.len()
    );

    // Get credentials for the account
    let storage = CredentialStorage::new()?;

    // Try to get stored session first (for OAuth accounts)
    let session = if let Some(stored_session) = storage.get_session(&react_args.reactAs)? {
        debug!("Using stored session for {}", react_args.reactAs);
        stored_session
    } else {
        // Fallback to creating new session with credentials (for app password accounts)
        debug!(
            "No valid session for {}, using app password flow",
            react_args.reactAs
        );
        let credentials = storage.get_credentials(&react_args.reactAs)?;
        let session_manager = SessionManager::new()?;
        session_manager.login(&credentials).await?
    };

    debug!("Authenticated as {} (DID: {})", session.handle, session.did);

    let client = crate::http::client_with_timeout(std::time::Duration::from_secs(120));

    // Track results
    let mut results = Vec::new();
    let mut errors = Vec::new();

    // Process likes
    for post_uri in &react_args.like {
        match process_like(&client, &session, post_uri).await {
            Ok(msg) => results.push(msg),
            Err(e) => errors.push(format!("Like failed for {}: {}", post_uri, e)),
        }
    }

    // Process unlikes
    for post_uri in &react_args.unlike {
        match process_unlike(&client, &session, post_uri).await {
            Ok(msg) => results.push(msg),
            Err(e) => errors.push(format!("Unlike failed for {}: {}", post_uri, e)),
        }
    }

    // Process reposts
    for post_uri in &react_args.repost {
        match process_repost(&client, &session, post_uri).await {
            Ok(msg) => results.push(msg),
            Err(e) => errors.push(format!("Repost failed for {}: {}", post_uri, e)),
        }
    }

    // Process deletes
    for post_uri in &react_args.delete {
        match process_delete(&client, &session, post_uri).await {
            Ok(msg) => results.push(msg),
            Err(e) => errors.push(format!("Delete failed for {}: {}", post_uri, e)),
        }
    }

    // Format results as markdown
    let mut markdown = String::from("# React Operations Results\n\n");

    if !results.is_empty() {
        markdown.push_str("## Successful Operations\n\n");
        for (i, result) in results.iter().enumerate() {
            markdown.push_str(&format!("{}. {}\n", i + 1, result));
        }
        markdown.push('\n');
    }

    if !errors.is_empty() {
        markdown.push_str("## Failed Operations\n\n");
        for (i, error) in errors.iter().enumerate() {
            markdown.push_str(&format!("{}. {}\n", i + 1, error));
        }
        markdown.push('\n');
    }

    markdown.push_str(&format!(
        "**Summary:** {} successful, {} failed\n",
        results.len(),
        errors.len()
    ));

    debug!(
        "React operations completed: {} successful, {} failed",
        results.len(),
        errors.len()
    );

    Ok(ToolResult::text(markdown))
}

/// Process a like operation
async fn process_like(
    client: &reqwest::Client,
    session: &crate::auth::Session,
    post_uri: &str,
) -> Result<String, AppError> {
    let (did, _rkey, uri, cid) = fetch_post_info(client, session, post_uri).await?;

    let url = format!("{}/xrpc/com.atproto.repo.createRecord", session.service);

    let body = serde_json::json!({
        "repo": session.did,
        "collection": "app.bsky.feed.like",
        "record": {
            "$type": "app.bsky.feed.like",
            "subject": {
                "uri": uri,
                "cid": cid
            },
            "createdAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        }
    });

    let response = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .json(&body)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Like request failed: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Like failed with status {}: {}",
            status, error_text
        )));
    }

    Ok(format!("Liked post: {} ({})", post_uri, did))
}

/// Process an unlike operation
async fn process_unlike(
    client: &reqwest::Client,
    session: &crate::auth::Session,
    post_uri: &str,
) -> Result<String, AppError> {
    // First, we need to find the like record for this post
    let (did, _rkey, uri, _cid) = fetch_post_info(client, session, post_uri).await?;

    // List likes to find the one for this post
    // Note: This lists up to 100 likes. For users with more likes, this is a known limitation.
    // A future improvement would be to implement pagination if the like is not found in the first page.
    let list_url = format!(
        "{}/xrpc/com.atproto.repo.listRecords?repo={}&collection=app.bsky.feed.like&limit=100",
        session.service, session.did
    );

    let response = client
        .get(&list_url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Failed to list likes: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        return Err(AppError::NetworkError(format!(
            "Failed to list likes: {}",
            status
        )));
    }

    let list_result: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::ParseError(format!("Failed to parse likes list: {}", e)))?;

    // Find the like record for this post
    let like_rkey = list_result["records"]
        .as_array()
        .and_then(|records| {
            records.iter().find_map(|record| {
                let subject_uri = record["value"]["subject"]["uri"].as_str()?;
                if subject_uri == uri {
                    record["uri"].as_str().and_then(|u| {
                        // Extract rkey from at://{did}/app.bsky.feed.like/{rkey}
                        u.split('/').next_back().map(|s| s.to_string())
                    })
                } else {
                    None
                }
            })
        })
        .ok_or_else(|| AppError::NotFound(format!("No like found for post: {}", post_uri)))?;

    // Delete the like record
    let delete_url = format!("{}/xrpc/com.atproto.repo.deleteRecord", session.service);

    let delete_body = serde_json::json!({
        "repo": session.did,
        "collection": "app.bsky.feed.like",
        "rkey": like_rkey
    });

    let response = client
        .post(&delete_url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .json(&delete_body)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Unlike request failed: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Unlike failed with status {}: {}",
            status, error_text
        )));
    }

    Ok(format!("Unliked post: {} ({})", post_uri, did))
}

/// Process a repost operation
async fn process_repost(
    client: &reqwest::Client,
    session: &crate::auth::Session,
    post_uri: &str,
) -> Result<String, AppError> {
    let (did, _rkey, uri, cid) = fetch_post_info(client, session, post_uri).await?;

    let url = format!("{}/xrpc/com.atproto.repo.createRecord", session.service);

    let body = serde_json::json!({
        "repo": session.did,
        "collection": "app.bsky.feed.repost",
        "record": {
            "$type": "app.bsky.feed.repost",
            "subject": {
                "uri": uri,
                "cid": cid
            },
            "createdAt": chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        }
    });

    let response = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .json(&body)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Repost request failed: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Repost failed with status {}: {}",
            status, error_text
        )));
    }

    Ok(format!("Reposted post: {} ({})", post_uri, did))
}

/// Process a delete operation
async fn process_delete(
    client: &reqwest::Client,
    session: &crate::auth::Session,
    post_uri: &str,
) -> Result<String, AppError> {
    let (did, rkey, _uri, _cid) = fetch_post_info(client, session, post_uri).await?;

    // Verify the post belongs to the authenticated user
    if did != session.did {
        return Err(AppError::InvalidInput(format!(
            "Cannot delete post that doesn't belong to you: {}",
            post_uri
        )));
    }

    let url = format!("{}/xrpc/com.atproto.repo.deleteRecord", session.service);

    let body = serde_json::json!({
        "repo": session.did,
        "collection": "app.bsky.feed.post",
        "rkey": rkey
    });

    let response = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .json(&body)
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Delete request failed: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Delete failed with status {}: {}",
            status, error_text
        )));
    }

    Ok(format!("Deleted post: {}", post_uri))
}

/// Fetch post information (DID, rkey, URI, CID) from a URI/URL
async fn fetch_post_info(
    client: &reqwest::Client,
    session: &crate::auth::Session,
    post_uri: &str,
) -> Result<(String, String, String, String), AppError> {
    // Parse the URI using the shared utility
    let post_ref = crate::bluesky::uri::parse_post_uri(post_uri).await?;

    // Fetch the post to get its CID
    let url = format!(
        "{}/xrpc/com.atproto.repo.getRecord?repo={}&collection=app.bsky.feed.post&rkey={}",
        session.service, post_ref.did, post_ref.rkey
    );

    let response = client
        .get(&url)
        .header("Authorization", format!("Bearer {}", session.access_jwt))
        .send()
        .await
        .map_err(|e| AppError::NetworkError(format!("Failed to fetch post: {}", e)))?;

    let status = response.status();
    if !status.is_success() {
        let error_text = response
            .text()
            .await
            .unwrap_or_else(|_| "Unknown error".to_string());
        return Err(AppError::NetworkError(format!(
            "Failed to fetch post with status {}: {}",
            status, error_text
        )));
    }

    let post_data: serde_json::Value = response
        .json()
        .await
        .map_err(|e| AppError::ParseError(format!("Failed to parse post data: {}", e)))?;

    let uri = post_data["uri"]
        .as_str()
        .ok_or_else(|| AppError::ParseError("No URI in post data".to_string()))?
        .to_string();
    let cid = post_data["cid"]
        .as_str()
        .ok_or_else(|| AppError::ParseError("No CID in post data".to_string()))?
        .to_string();

    Ok((post_ref.did, post_ref.rkey, uri, cid))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_react_args_parsing() {
        let args = json!({
            "reactAs": "test.bsky.social",
            "like": ["at://did:plc:abc/app.bsky.feed.post/123"],
            "unlike": [],
            "repost": [],
            "delete": []
        });

        let parsed: ReactArgs = serde_json::from_value(args).unwrap();
        assert_eq!(parsed.reactAs, "test.bsky.social");
        assert_eq!(parsed.like.len(), 1);
        assert_eq!(parsed.unlike.len(), 0);
        assert_eq!(parsed.repost.len(), 0);
        assert_eq!(parsed.delete.len(), 0);
    }

    #[tokio::test]
    async fn test_react_args_mixed_operations() {
        let args = json!({
            "reactAs": "test.bsky.social",
            "like": ["at://did:plc:abc/app.bsky.feed.post/1"],
            "unlike": ["at://did:plc:abc/app.bsky.feed.post/2"],
            "repost": ["at://did:plc:def/app.bsky.feed.post/3"],
            "delete": ["at://did:plc:ghi/app.bsky.feed.post/4"]
        });

        let parsed: ReactArgs = serde_json::from_value(args).unwrap();
        assert_eq!(parsed.reactAs, "test.bsky.social");
        assert_eq!(parsed.like.len(), 1);
        assert_eq!(parsed.unlike.len(), 1);
        assert_eq!(parsed.repost.len(), 1);
        assert_eq!(parsed.delete.len(), 1);
    }
}
